### PR TITLE
revision: default templates, structure

### PR DESCRIPTION
**Pull request by omikron-fp**
Quite a bit. Still contains some TODOs.

#### Proposed change of element name to `ff-filter-cloud`.
`ff-asn-selected-filter-cloud` is quite bulky.
Generic name should make it obvious to be an independent element that can be placed anywhere in the DOM.
If the ASN gets a rewrite as e.g. `ff-facets`, the filter cloud element could serve that element, too.
If the `asn` part in the name is required, `ff-asn-filter-cloud` might also work.

#### slot="unselected" unclear
Not quite clear what a possible use case could be. Filter cloud only displays selected filters anyway.

#### ff-asn-group-element unclear
Similar to `slot="unselected"`.
In case `asn` is dropped from the name, `ff-asn-group-element` would be out of place.

#### Remove all
The "remove all" functionality could be offered built-in. Alternatively, an example incorporating the existing element could spark interest.

### CODE REVIEW

File: specs/ff-asn-selected-filter-cloud/ff-asn-selected-filter-cloud.md [hunk 1]
```diff
@@ -1,86 +1,212 @@
...
-        <!-- Will be rendered inside this section -->
-        <ff-asn-group-element>
-            <span slot="selected">{{name}}</span>
-            <span slot="unselected">{{name}}</span>
-        </ff-asn-group-element>
-    </section>
-</ff-asn-selected-filter-cloud>
+```html
+<ff-filter-cloud>
+    <h4>Teh Filter Cloud™</h4>
```
Comment (ghost) [inline specs/ff-asn-selected-filter-cloud/ff-asn-selected-filter-cloud.md]:
```suggestion
    <h4>The Filter Cloud</h4>
```

File: specs/ff-asn-selected-filter-cloud/ff-asn-selected-filter-cloud.md [hunk 2]
```diff
@@ -1,86 +1,212 @@
...
 ```html
-<ff-asn-selected-filter-cloud>
-    <ff-asn-group-element>
-        <span slot="selected">{{name}}</span>
-        <span slot="unselected">{{name}}</span>
-    </ff-asn-group-element>
-</ff-asn-selected-filter-cloud>
+<ff-filter-cloud>
+    <div data-filter-container>
+        <span data-filter>{{name}}</span>  <!-- this element is repeated inside the filter-container -->
```
Comment (ghost) [inline specs/ff-asn-selected-filter-cloud/ff-asn-selected-filter-cloud.md]:
```suggestion
        <span data-filter>{{name}}</span>  <!-- this template will be rendered for each filter inside the data-filter-container -->
```

File: specs/ff-asn-selected-filter-cloud/ff-asn-selected-filter-cloud.md [hunk 3]
```diff
@@ -1,86 +1,212 @@
...
+</ff-filter-cloud>
+```
+
+
+## Structural attributes
+
+| Attribute             | Required | Description
+| ---------             | -------- | -----------
+| data-filter-container | optional | (TODO) Warning in console if not specified
+| data-filter           | optiional| (TODO) Warning in console if not specified
```
Comment (ghost) [inline specs/ff-asn-selected-filter-cloud/ff-asn-selected-filter-cloud.md]:
```suggestion
| data-filter           | optional| (TODO) Warning in console if not specified
```

**Comment by toarm** [comment]
> Not quite clear what a possible use case could be. Filter cloud only displays selected filters anyway.

`ff-asn-group-element` switches to unselected state immediately after clicking to reflect the desired state. In addition the element knows already how to deselect a filter.

edit:
In general the intent of using `ff-asn-group-element` is composition and resuing our own elements and code. If we consider a rewrite (ff-facets or similar) it might be worth to exclude it.

File: specs/ff-asn-selected-filter-cloud/ff-asn-selected-filter-cloud.md [hunk 4]
```diff
@@ -1,86 +1,212 @@
...
+        <span data-filter>
+            <span>Custom: Filter n</span>
+            <span deselect>×</span>
+        </span>
+    </div>
+</ff-filter-cloud>
+```
+
+
+## Structural attributes
```
Comment (ghost) [inline specs/ff-asn-selected-filter-cloud/ff-asn-selected-filter-cloud.md]: I am not too happy with the term _Structural attributes_, as it is too broad for me. Maybe something like _Template specifier_ (these attributes are the hooks for our template engines, aren't they?)

File: specs/ff-asn-selected-filter-cloud/ff-asn-selected-filter-cloud.md [hunk 5]
```diff
@@ -1,86 +1,212 @@
...
+
+### Attributes
+
+| Attribute | Required | Type   | Options        | Default     | JS property
+| --------- | -------- | ----   | -------        | -------     | -----------
+| order     | optional | String | _fact-finder_  | fact-finder | `order`
+|           |          |        | alphabetical   |             |
+|           |          |        | user-selection |             |
+
+#### Options
```
Comment (ghost) [inline specs/ff-asn-selected-filter-cloud/ff-asn-selected-filter-cloud.md:153]: Could we intend this, to make it more clear, that it refers to the attribute `order`?

**Review by ghost — COMMENTED** [review]
Can I intrigue you, to consider using `<template>` elements for our template? That would
	- remove the danger of FOUC out of the box
	- remove the need to remove the template from DOM
	- be a little more performant for free
	
In addition I would loved to see a naming convention like `ff-template-role="deselected"`, `ff-template-role="data-filter"`... Would make the mapping easier of which parts gets reused by our template engine. In the very long term a common approach would also stop the bug source of when to use `data-container`, `data-someName`, `slot` or a specific attribute name, which is likely when changing from `1.x` to `3.x`. As something like that would have a clear and easy to understand name schema, I assume it doesn't even hurt in the short term.

**Comment by ghost** [comment]
> In general the intent of using `ff-asn-group-element` is composition and resuing our own elements and code. If we consider a rewrite (ff-facets or similar) it might be worth to exclude it.

In general I love the idea. But I would recommend to reuse smaller components. It shouldn't be too hard to extract the logic of deselecting a filter into some reusable component / function and reuse it in many places. As this would only deselect the filter, it is independent of things like asn, facet or whatever might change :)

File: specs/ff-asn-selected-filter-cloud/ff-asn-selected-filter-cloud.md (hunk 4 above)
Comment (toarm) [reply specs/ff-asn-selected-filter-cloud/ff-asn-selected-filter-cloud.md]: Dito, we currently call it just `properties` everywhere. I dont like to introduce a new naming just for one element
edit:
missunderstood this one x/
In the documentation we call it `directive`

File: specs/ff-asn-selected-filter-cloud/ff-asn-selected-filter-cloud.md [hunk 6]
```diff
@@ -1,86 +1,212 @@
...
-            <span slot="selected">{{name}}</span>
-            <span slot="unselected">{{name}}</span>
-        </ff-asn-group-element>
-    </section>
-</ff-asn-selected-filter-cloud>
+```html
+<ff-filter-cloud>
+    <h4>The Filter Cloud</h4>
+
+    <div data-filter-container>
```
Comment (toarm) [inline specs/ff-asn-selected-filter-cloud/ff-asn-selected-filter-cloud.md]: please remove the `data-filter-container` attribute. It's actually unnecessary.

File: specs/ff-asn-selected-filter-cloud/ff-asn-selected-filter-cloud.md (hunk 4 above)
Comment (toarm) [reply specs/ff-asn-selected-filter-cloud/ff-asn-selected-filter-cloud.md]: See above:
> In the documentation we call it `directive`

File: specs/ff-asn-selected-filter-cloud/ff-asn-selected-filter-cloud.md [hunk 7]
```diff
@@ -1,86 +1,212 @@
-# ff-asn-selected-filter-cloud
-The `ff-asn-selected-filter-cloud` displays all active filters using `ff-asn-group-element`'s.
+# ff-filter-cloud
```
Comment (toarm) [inline specs/ff-asn-selected-filter-cloud/ff-asn-selected-filter-cloud.md:1]: What do you think about `selected-filter-cloud`? we have an element called `ff-tagcloud` in 1.x. I think it's a bit missleading if we just call it `filter-cloud`

**Review by toarm — COMMENTED** [review]
I'm still not convinced about using `<some-element data-filter>` over `<ff-asn-group-element>`. We already have all the funcitonality we need in the group-element.

File: specs/ff-asn-selected-filter-cloud/ff-asn-selected-filter-cloud.md (hunk 6 above)
Comment (omikron-fp) [reply specs/ff-asn-selected-filter-cloud/ff-asn-selected-filter-cloud.md]: How would you decide where to put the filter elements?
Putting them directly into `ff-filter-cloud` prevents adding further HTML after the elements.
Something like this shouldn't be possible:
```html
<ff-filter-cloud>
    <h4>The Filter Cloud</h4>
    <div data-filter-container></div>
    <h4>Footer</h4>
</ff-filter-cloud>
```

**Comment by omikron-fp** [comment]
> I'm still not convinced about using `<some-element data-filter>` over `<ff-asn-group-element>`. We already have all the funcitonality we need in the group-element.

Doesn't it actually contain more functionality than we need? Like the `unselected` slot. I'm worried that unneeded functionality could cause confusion or at least wrong expectations.

And it would tighly couple the new element to `ff-asn`. Is creating a generic element potentially serving other asn-like elements not a concern?

**Pull request merged**